Enabling custom TF signature draft

dimitreOliveira opened the pull request:
# What does this PR do?
This PR is related to the discussion on issue #19094 , which is related to `transformers` enabling users to provide custom signatures for models while saving them with [save_pretrained](https://github.com/huggingface/transformers/blob/ca485e562b675341409e3e27724072fb11e10af7/src/transformers/modeling_tf_utils.py#L2085).

Fixes #19094 


## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] Did you write any new necessary tests?

dimitreOliveira committed:
Custom TF signature draft

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_19249). All of your documentation changes will be reflected on that endpoint.

gante commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -2166,8 +2169,10 @@ def save_pretrained(
             files_timestamps = self._get_files_timestamps(save_directory)
 
         if saved_model:
+            if not signatures:
```
```suggestion
            if signatures is not None:
```
(same thing in practice, but might help catching incorrect falsy values in `signatures`)

gante reviewed (approved):
Thank you for adding this feature 👍

gante commented:
Adding a test (perhaps inside [this test class](https://github.com/huggingface/transformers/blob/f3d2f7a6e08efe18debf59512325f02128394b43/tests/test_modeling_tf_common.py#L1913)) would be great!

Rocketknight1 commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -2136,6 +2137,8 @@ def save_pretrained(
                 repository you want to push to with `repo_id` (will default to the name of `save_directory` in your
                 namespace).
+            signatures (`dict` or `tf.function`, *optional*, defaults to `None`):
+                Model's signature used for serving.
```
```suggestion
                Model's signature used for serving. This will be passed to the `signatures` argument of model.save().
```

Rocketknight1 reviewed (approved):
Looks good to me! I made a small suggestion just to clarify for users where exactly the signatures arg is going.

sgugger commented on src/transformers/modeling_tf_utils.py:
```diff
@@ -2136,6 +2137,8 @@ def save_pretrained(
                 Whether or not to push your model to the Hugging Face model hub after saving it. You can specify the
                 repository you want to push to with `repo_id` (will default to the name of `save_directory` in your
                 namespace).
+            signatures (`dict` or `tf.function`, *optional*, defaults to `None`):
```
```suggestion
            signatures (`dict` or `tf.function`, *optional*):
```
We don't indicate the default when it's None.

sgugger reviewed (approved):
Thanks for your PR!

dimitreOliveira committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Matt <Rocketknight1@users.noreply.github.com>
Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

dimitreOliveira committed:
Adding tf signature tests

dimitreOliveira committed:
Fixing signature check and adding asserts

dimitreOliveira committed:
fixing model load path

dimitreOliveira commented:
Hey there, I think that this code is finished and ready to be merged, I fixed the code and added the tests, let me know if the tests look too hardcoded or if I could check something else.

cc @gante @Rocketknight1 @sgugger

gante commented on tests/test_modeling_tf_common.py:
```diff
@@ -2190,6 +2190,39 @@ def test_checkpoint_sharding_local(self):
+    def test_save_pretrained_signatures(self):
+        model = TFBertModel.from_pretrained("hf-internal-testing/tiny-random-bert")
+
+        # Short custom TF signature function.
+        # `input_signature` is specific to BERT.
+        @tf.function(
+            input_signature=[
+                [
+                    tf.TensorSpec([None, None], tf.int32, name="input_ids"),
+                    tf.TensorSpec([None, None], tf.int32, name="token_type_ids"),
+                    tf.TensorSpec([None, None], tf.int32, name="attention_mask"),
+                ]
+            ]
+        )
+        def serving_fn(input):
+            return model(input)
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Using default signature (default behavior)
+            model.save_pretrained(tmp_dir, saved_model=True, signatures=None)
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
+
+            # Providing custom signature function
+            model.save_pretrained(tmp_dir, saved_model=True, signatures=serving_fn)
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
+
+            # Providing custom signature function (dict input)
+            model.save_pretrained(tmp_dir, saved_model=True, signatures={"serving_default": serving_fn})
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
```
There are a few issues with the current test :p 
1. The test asserts that there is a signature, but at some point in the future, we might remove the logic related to the argument you added in this PR (`signatures`) and the tests would still pass because the default signature would be loaded. We should test that we get a different signature when we pass a signature. For instance, the new signature could have two inputs (`input_ids` and `attention_mask`), and we could test the signature lengths are `3`, `2`, and `2`, following your test order.
2. I'm unsure whether saving a model into the same folder will store the model in a subfolder ending in `/1`, or if there might be other conflicting behavior. To be safe, add a `with tempfile.TemporaryDirectory() as tmp_dir:` context before each test, to ensure it stores the model on a new folder.

dimitreOliveira replied:
Hey there @gante , thanks for the suggestions, I made a few updates to address them, let me know if it looks better now.

Regarding your points:
>1. The test asserts that there is a signature, but at some point in the future, we might remove the logic related to the argument you added in this PR (signatures) and the tests would still pass because the default signature would be loaded. We should test that we get a different signature when we pass a signature. For instance, the new signature could have two inputs (input_ids and attention_mask), and we could test the signature lengths are 3, 2, and 2, following your test order.
This is a great point, as you can see I updated the tests to check for the custom names of the signatures, I am not aware of a way to check the input names of the signature as you suggested, I can only check the signature names if you know how to do it let me know and I can update the tests.

>2. I'm unsure whether saving a model into the same folder will store the model in a subfolder ending in /1, or if there might be other conflicting behavior. To be safe, add a with tempfile.TemporaryDirectory() as tmp_dir: context before each test, to ensure it stores the model on a new folder.
Actually, the subsequent model savings will subscribe the previous ones, but your suggestion is pretty good since it makes the tests cleaner, thanks!

gante reviewed (commented):
A few nits on the test :)

Commit:
Adjusting signature tests

dimitreOliveira committed:
Formatting file

gante commented:
@dimitreOliveira it seems ready to merge.

There is an apparently unrelated failure, but let's rule it out. Can I request you to rebase and commit? i.e. ensure your fork is up to date with the latest main and then run
```bash
git checkout main
git pull
git checkout custom_tf_signature
git rebase origin/main
git push origin custom_tf_signature -f
```

dimitreOliveira commented:
@gante should be good now, thanks!
